_posts/2023-01-14-trivially-relocatable.md

```diff
@@ -0,0 +1,656 @@
+`memcpy(&s, &from, sus::mem::data_size_of<S>())`, then we copy only 5 bytes (the `int32_t a`
+and `int8_t b`) into `s` and avoid clobbering the members of `T` or any other subclass in its tail
+padding.
+
+##### The `[[no_unique_address]]` attribute
+
+C++20 introduces the `[[no_unique_address]]` attribute which can appear on a class member
+declaration. It tells the compiler to allow the member's tail padding to be used by later
+members.
+
+This new attribute allows us to compose a new type with a member `S` and which makes use of the tail
+padding in `S`. Here we use the same non-Standard-Layout type `S` from the previous section.
+We add another member below `S` but the size of `T`
+[is again no larger than S](https://godbolt.org/z/GhbvTaecM) because the
+member `T::c` has been located in the tail padding of `S`.
+
+```cpp
+struct T {
+    [[no_unique_address]] S s;
+    int8_t c;
+    // 2 bytes of tail padding.
+};
+static_assert(sizeof(S) == 8);
+static_assert(sizeof(T) == 8);
+```
+
+The side effect of the `[[no_unique_address]]` attribute is that `memcpy()` can do the wrong thing
+again. If we `memcpy(&s, &from, sizeof(S))` into an `S*` that points to the member `T::s` of `T`,
+we will copy the tail padding of `from` into `T::c`, clobbering its value with garbage. Here again,
+if we `memcpy(&s, &from, sus::mem::data_size_of<S>())` then we copy only 5 bytes (the `int32_t a`
+and `int8_t b`) into `s` and avoid clobbering anything in its tail padding.
+
+Since this behaviour is all implementation defined, we have unfortunate differences between
+compilers however, which can make testing for behaviour tricky. As in the
+[base class example](#base-classes), MSVC 19
+[does not](https://godbolt.org/z/TxoP4h9jY) make use of the tail padding in `S`
+and the size of `T` will be 12 as a result. This [also holds](https://godbolt.org/z/x6szhYPxo) when
+using the compiler-specific [`[[msvc::no_unique_address]]` attribute](
+https://devblogs.microsoft.com/cppblog/msvc-cpp20-and-the-std-cpp20-switch/) instead.
+
+##### Sorry but memcpy() with sizeof(T) is dangerous
+
+The outcome of the above is that `memcpy(dest, src, sizeof(T))` is dangerous in generic code, and
+even more so in C++20. The more correct thing to do is
+`memcpy(dest, src, sus::mem::data_size_of<T>())`.
+
+#### The Limits of Data size in a Library
+
+For some types, we can not determine a data size. In particular, a union's data size is dynamic
+depending on its active member. If we could enumerate the members of a union, we could use the
+maximum data size of all its members, but that is beyond the scope of what a library can achieve
+unfortunately.
+
+For that reason, the implementation of `sus::mem::data_size_of<T>()` on a union type
+[will return 0](https://github.com/chromium/subspace/blob/21a55214fdd968ba01118697721b708b83540521/subspace/mem/__private/data_size_finder.h#L61-L65).
+This means that a data size of 0 should be treated as unknown, and this explains why we require

```

Comment:
"A data size of 0 is treated as an unknown data size, and this is why `sus::mem::...` requires a non-zero data size."

Reply:
"And since a data size of 0 is returned for an unknown data size, `sus::mem::relocate_by_memcpy`
requires a non-zero data size." 👍